### PR TITLE
TandemFoil: 5L/256d architecture at lr=2e-4 full budget

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
PR #2595 showed 5L/256d achieving -30.1% improvement at lr=3e-4, but that run suffered epoch starvation (too few epochs to converge). With the correct lr=2e-4 (established as optimal for TandemFoil) and the full 180-minute budget, 5L/256d should outperform both 3L/192d and 4L/256d. Deeper networks have more representational capacity for complex pressure distributions around tandem foil configurations.

## Instructions
Run the TandemFoil command with 5L/256d architecture at the correct lr=2e-4:

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 2e-4 \
  --cosine-t-max 10 \
  --no-use-ema \
  --model-slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --wandb_group tandem-arch-scaling \
  --epochs 9999
```

Changes vs baseline:
- `--model-layers 5` (was 3)
- `--model-hidden-dim 256` (was 192)
- `--model-heads 4` (was default)
- `--lr 2e-4` (was 3e-4 in PR #2595)

## Baseline
Current best TandemFoil metrics (PR #2610, W&B: ixs1rqgk):
- val_primary/surface_pressure_mae: **45.07** (lower is better)

Reproduce baseline:
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kakashi:

## Results

### Key Metrics

| Metric | Baseline (5L/256d lr=3e-4, PR #2595) | PR baseline (PR #2610) | This run (5L/256d lr=2e-4) | vs PR #2595 |
|--------|------|------|------|------|
| **val_primary/surface_pressure_mae** | 52.81 | 45.07 | **65.23** | **+23.5% (worse)** |
| **test_primary/surface_pressure_mae** | 55.25 | — | **78.72*** | **+42.5% (worse)** |

*Test evaluated on the corrupted model at end of training (epoch 87), not on best val checkpoint (epoch 35). See analysis below.

### Per-Split Test Surface Pressure MAE (from corrupted model)

| Split | Baseline (PR #2595) | This run |
|-------|----------|----------|
| single_in_dist | 61.80 | 84.34 |
| geom_camber_rc | 60.20 | 90.04 |
| geom_camber_cruise | 48.99 | 68.10 |
| re_rand | 50.00 | 72.41 |

### Training Instability Timeline

| Epoch | Event | val_primary | Grad Norm |
|-------|-------|-------------|-----------|
| 16 | Pre-spike best | 70.29 | 1.23 |
| 23 | First spike | 104.45 | 2.33 |
| 35 | **Best val** | **65.23** | 1.10 |
| 36 | **GRADIENT EXPLOSION** | 124.56 | **1162.26** |
| 37 | Post-explosion | 100.66 | 1.90 |
| 40 | NaN grad norm | 79.35 | **NaN** |
| 40-87 | **Val frozen** | 79.35 | ~1.13 |

### Run Details
- **Command**: `cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py --dataset tandemfoil --optimizer lion --lr 2e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --model-layers 5 --model-hidden-dim 256 --model-heads 4 --epochs 9999 --agent kakashi --wandb-name "kakashi/tandem-5L256d-lr2e4" --wandb-group tandem-arch-scaling`
- **W&B run**: `lcqrlorc`
- **Epochs**: 87 (180-min timeout)
- **Best epoch**: 35 (val 65.23)

### What happened

**Negative result.** 5L/256d at lr=2e-4 significantly underperforms the same architecture at lr=3e-4 (PR #2595: 52.81 val). Key findings:

1. **lr=2e-4 is too low for 5L/256d with Lion**: Convergence was much slower — best val of 65.23 reached at epoch 35 vs the 5L baseline achieving 52.81 at epoch 67 with lr=3e-4. Even with more epochs available, this run was on a worse trajectory.

2. **Catastrophic gradient explosion at epoch 36**: Grad norm spiked from 1.10 to **1162.26** in a single epoch — a 1000x increase. This destroyed model weights. Without gradient clipping, there is no defense against such events.

3. **Model permanently corrupted**: After a NaN grad norm at epoch 40, validation metrics froze at 79.35 for the remaining 47 epochs (40-87). Training loss continued to change but produced no useful model updates. ~94 minutes of compute wasted on a dead model.

4. **Direct comparison with PR #2711 (6L/256d + grad-clip=1.0)**: That run achieved **40.78 val** with zero gradient explosions. Even the 5L baseline at lr=3e-4 (52.81) significantly beats this run. The gradient clipping story is now very clear:
   - With grad-clip=1.0: smooth convergence, best val 40.78, still improving at cutoff
   - Without grad-clip: gradient explosion at epoch 36, model destroyed, best val 65.23

5. **Baseline note**: PR body lists 45.07 (PR #2610) as baseline, but BASELINE.md shows 52.81 (PR #2595) and our PR #2711 achieved 40.78 — the PR baseline appears stale.

### Suggested follow-ups
1. **Do NOT use lr=2e-4 for TandemFoil with Lion** — lr=3e-4 is clearly better at this architecture scale.
2. **Gradient clipping is essential**: Every TandemFoil run should include `--grad-clip 1.0` (or similar). The gradient explosion evidence is now very strong across multiple runs.
3. **Investigate the frozen model bug**: The val metrics freezing after NaN gradients suggests a potential bug in model evaluation — perhaps NaN weights cause deterministic outputs. Consider adding NaN detection/early stopping as a safeguard.
4. **Focus on 6L+ with grad-clip**: The 6L/256d + grad-clip=1.0 result (40.78) is the clear frontier for TandemFoil. Further depth scaling with clipping is the most promising direction.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2775 because ADVISOR: Closing as broken run. W&B confirms val=65.23 vs 44.72 baseline — 46% above baseline. Gradient explosion at epoch 36 (grad norm 1.10 → 1162.26) permanently corrupted the model, wasting 94 minutes of compute. The 5L/256d architecture for TandemFoil is inherently unstable without gradient clipping at lr=2e-4. This direction should only be revisited with --grad-clip enabled to prevent the explosion. Thank you for the thorough diagnostics — the grad norm data is valuable for calibrating future TandemFoil experiments..
